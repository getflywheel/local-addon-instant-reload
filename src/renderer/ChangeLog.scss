@import '../../../../node_modules/@getflywheel/local-components/src/styles/_partials/index.scss';
@import '../../../_assets/fonts/JetBrainsMono-Regular.ttf';

.sessionLogContainer {
	transition: box-shadow .1s;
	outline: none;
	border: 0;
	border-radius: 4px;
	padding: 10px;
	width: 100%;
	background-color: $gray1;
	@include theme-input-border-box-shadow;
	margin-bottom: 15px;
	font-size: 11px;
	line-height: 1.91;
	max-height: 214px;
	overflow-y: auto;
	word-wrap: break-word;
	font-family: 'JetBrainsMono', sans-serif;

	@include if-theme-dark() {
		background-color: $gray-dark !important;
	}
}
