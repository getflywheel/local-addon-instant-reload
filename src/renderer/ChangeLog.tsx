import React from 'react';
import type { FileChangeEntry } from '../types';

// import styles from './ChangeLog.scss';

/**
 * @todo remove the stubbed styled
 */
const styles = {};

interface Props {
	changeLog? : FileChangeEntry[];
}

export const sessionLogPlaceholderCopy = 'Start editing to see changes logged here!';

function formatLogItem (logItem: FileChangeEntry, isLastItem: boolean): string {
	const filePathRegExp = new RegExp(/[^]*\/app\/public/);
	const {
		timeChanged,
		eventType,
		fileName,
		fileSize,
	} = logItem;

	let verb = 'Changed';
	let size = `(${fileSize})`;

	if (['unlink', 'unlinkDir'].includes(eventType)) {
		verb = 'Deleted';
		size = '';
	} else if (['add', 'addDir'].includes(eventType)) {
		verb = 'Added';
	}

	return `${isLastItem ? '>' : ''}[${timeChanged}] - ${verb} - ${fileName.replace(filePathRegExp, '')} ${size}`;
}

/* eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types */
const InstantReloadSessionLog = ({ changeLog = [] }: Props) => {
	const displayPlaceHolder = changeLog.length === 0;
	console.log('change log component', changeLog)

	if (displayPlaceHolder) {
		return (
			<li>{sessionLogPlaceholderCopy}</li>
		);
	}

	return (
		<div className={styles.sessionLogContainer}>
			<ul>
				{changeLog.map((logItem: FileChangeEntry, i: number) => (
					/**
					 * disable this rule because there is a slight chance that time stamps will match and produce
					 * elements with identical keys. Using the array index will actually help to ensure unique keys
					 * in this case.
					 */
					/* eslint-disable-next-line react/no-array-index-key */
					<li key={`${i}-${logItem.timeChanged}-${logItem.fileName}-${logItem.fileSize}`}>
						{formatLogItem(logItem, i === changeLog.length - 1)}
					</li>
				))}
			</ul>
		</div>
	);
};

export default InstantReloadSessionLog;
