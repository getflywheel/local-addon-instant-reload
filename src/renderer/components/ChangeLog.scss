@import '../../../node_modules/@getflywheel/local-components/src/styles/_partials/index.scss';

.sessionLogContainer {
	transition: box-shadow .1s;
	outline: none;
	border: 0;
	border-radius: 4px;
	padding: 10px;
	width: 100%;
	background-color: $gray1;
	@include theme-input-border-box-shadow;
	margin-bottom: 15px;
	font-size: 11px;
	line-height: 1.91;
	max-height: 214px;
	overflow-y: auto;
	word-wrap: break-word;
	/** @todo ensure that this font will still load properly after stripping the font out of Local */
	font-family: 'JetBrainsMono', sans-serif;

	ul {
		list-style-type: none;
		padding: 0;
	}

	@include if-theme-dark() {
		background-color: $gray-dark !important;
	}
}
