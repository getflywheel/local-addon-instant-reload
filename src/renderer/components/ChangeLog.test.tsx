import React from 'react';
import { create, act } from 'react-test-renderer';
import ChangeLog, { sessionLogPlaceholderCopy } from './ChangeLog';
import type { FileChangeEntry } from '../../types';

const dummyLogItem: FileChangeEntry = {
	fileName: 'newfile.extension',
	eventType: 'add',
	timeChanged: '4:20:00 PM',
	fileSize: '121kb',
};

const commonProps = {
	changeLog: [dummyLogItem],
};

const makeTree = (props = {}) => {
	let tree;
	act(() => {
		tree = create(
			<ChangeLog {...props} />,
		);
	});
	return tree;
};

describe('ChangeLog', () => {
	it('renders contents correctly', () => {
		const tree = create(
			<ChangeLog {...commonProps} />,
		).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('renders empty array message correctly', () => {
		const tree = makeTree({ sessionLog: [] });
		expect(
			tree.root.findByType('ul').findByType('li').children,
		).toEqual([sessionLogPlaceholderCopy]);
	});

	it('correctly adds a log line', () => {
		const tree = makeTree(commonProps);
		let sessionLog = tree.root
			.findByType('ul')
			.findAllByType('li');

		expect(sessionLog.length).toBe(1);

		act(() => {
			tree.update(
				<ChangeLog
					changeLog={[
						{
							fileName: '/wp-content/themes/twentytwenty/style.css',
							timeChanged: '3:01:45 PM',
							fileSize: '121kb',
							eventType: 'change',
						},
						{
							fileName: '/wp-content/themes/twentytwenty/style.css',
							timeChanged: '4:20:00 PM',
							fileSize: '125kb',
							eventType: 'change',
						},
					]}
				/>,
			);
		});

		sessionLog = tree.root
			.findByType('ul')
			.findAllByType('li');

		expect(sessionLog.length).toBe(2);
	});
});
