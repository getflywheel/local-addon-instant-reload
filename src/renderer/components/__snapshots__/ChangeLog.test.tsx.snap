// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeLog renders contents correctly 1`] = `
<div
  className="sessionLogContainer"
>
  <ul>
    <li>
      &gt;[4:20:00 PM] - Added - newfile.extension (121kb)
    </li>
  </ul>
</div>
`;
